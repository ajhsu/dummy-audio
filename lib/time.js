export const seconds = 1000;
export const minutes = 60 * seconds;