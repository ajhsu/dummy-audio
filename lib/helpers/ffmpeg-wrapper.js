import lodash from 'lodash';
import { exec } from 'child_process';
import config from '../config';
import { seconds, minutes } from '../time';

// 使用 FFMPEG 建立靜音檔
export const createSilentAudio = ({ duration }) => {
  if (
    !lodash.isNumber(duration) ||
    !(duration > 0) ||
    duration > 10 * minutes
  ) {
    throw new Error('invalid duration');
  } else {
    const fileName = `silent-${duration}ms.mp3`;
    return {
      fileName,
      command: `ffmpeg -f lavfi -i aevalsrc=0:0::duration=${duration *
        0.001}:sample_rate=44100 -ab 128k ${config.localAudioCacheFolderName}/${fileName} -y`
    };
  }
};

// 用以執行外部指令
export const execute = command =>
  new Promise((y, n) => {
    exec(command, (error, stdout, stderr) => {
      if (error) {
        n(`exec error: ${error}`);
        return;
      }
      y({ stdout, stderr });
    });
  });
