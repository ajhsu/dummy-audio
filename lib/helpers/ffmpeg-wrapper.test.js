import { seconds, minutes } from '../time';
import { createSilentAudio, execute } from './ffmpeg-wrapper';

test('createSilentAudio', () => {
  {
    const silentAudio = createSilentAudio({ duration: 30 * seconds });
    expect(silentAudio.command).toBe(
      `ffmpeg -f lavfi -i aevalsrc=0:0::duration=30:sample_rate=44100 -ab 128k caches/silent-30000ms.mp3 -y`
    );
  }
  // 音檔長度上限十分鐘
  {
    expect(() => {
      createSilentAudio({ duration: 10 * minutes + 1 });
    }).toThrow();
  }
  {
    expect(() => {
      createSilentAudio({ duration: 'test' });
    }).toThrow();
  }
  {
    expect(() => {
      createSilentAudio();
    }).toThrow();
  }
  {
    expect(() => {
      createSilentAudio({ duration: 0 });
    }).toThrow();
  }
  {
    expect(() => {
      createSilentAudio({ duration: -1 });
    }).toThrow();
  }
  {
    expect(() => {
      createSilentAudio({ duration: NaN });
    }).toThrow();
  }
  {
    expect(() => {
      createSilentAudio({ duration: Number.MAX_VALUE });
    }).toThrow();
  }
});

test('execute', async () => {
  const result = await execute('echo "hello dummy.audio"');
  expect(result).toEqual({
    stdout: 'hello dummy.audio\n',
    stderr: ''
  });
});
