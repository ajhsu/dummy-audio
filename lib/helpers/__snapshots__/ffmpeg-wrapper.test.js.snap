// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSilentAudio 1`] = `"ffmpeg -f lavfi -i aevalsrc=0:0::duration=30:sample_rate=44100 -ab 128k caches/silent-30000ms.mp3 -y"`;
